File: app/src/main/java/com/pietrantuono/podcasts/addpodcast/customviews/PodcastsRecycler.java
package com.pietrantuono.podcasts.addpodcast.customviews;


import android.content.Context;
import android.content.res.Configuration;
import android.support.annotation.Nullable;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.AttributeSet;

import com.pietrantuono.podcasts.addpodcast.model.pojos.Podcast;
import com.pietrantuono.podcasts.application.App;

import java.util.List;

import javax.inject.Inject;

public class PodcastsRecycler extends RecyclerView {
    @Inject PodcastsAdapter adapter;

    public PodcastsRecycler(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    private void init() {
        ((App)getContext().getApplicationContext()).getApplicationComponent().inject(this);
        setLayoutManager(createLayoutManager());
        setAdapter(adapter);
    }

    private LayoutManager createLayoutManager() {
        int orientation = getResources().getConfiguration().orientation;
        if (orientation == Configuration.ORIENTATION_LANDSCAPE) {
            return new GridLayoutManager(getContext(), 4);
        } else {
            return new GridLayoutManager(getContext(), 3);
        }
    }

    public void setItems(List<Podcast> items) {
        adapter.setItems(items);
    }

    private void setOnItemClickListener(PodcastsAdapter.OnItemClickedClickedListener onItemClickedClickedListener) {
        adapter.setOnItemClickListener(onItemClickedClickedListener);
    }

    public void setListeners(PodcastsAdapter.OnItemClickedClickedListener addPodcastPresenter) {
        setOnItemClickListener(addPodcastPresenter);
    }

    public boolean isPartiallyHidden(int position) {
        return ((LinearLayoutManager) getLayoutManager()).findLastCompletelyVisibleItemPosition() < position;
    }
}

File: app/src/main/java/com/pietrantuono/podcasts/addpodcast/customviews/PodcastsRecycler.kt
package com.pietrantuono.podcasts.addpodcast.customviews


import android.content.Context
import android.content.res.Configuration
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.util.AttributeSet


import com.pietrantuono.podcasts.application.App
import models.pojos.Podcast

import javax.inject.Inject

class PodcastsRecycler(context: Context, attrs: AttributeSet?) : RecyclerView(context, attrs) {
    @Inject lateinit var adapter: PodcastsAdapter

    init {
        (context.applicationContext as App).appComponent?.inject(this)
        layoutManager = createLayoutManager()
        setAdapter(adapter)
    }

    private fun createLayoutManager(): RecyclerView.LayoutManager {
        if (resources.configuration.orientation == Configuration.ORIENTATION_LANDSCAPE) {
            return GridLayoutManager(context, 4)
        } else {
            return GridLayoutManager(context, 3)
        }
    }

    fun setItems(items: List<Podcast>) {
        adapter.setItems(items)
    }

    private fun setOnItemClickListener(onItemClickedClickedListener: PodcastsAdapter.OnItemClickedClickedListener) {
        adapter.setOnItemClickListener(onItemClickedClickedListener)
    }

    fun setListeners(addPodcastPresenter: PodcastsAdapter.OnItemClickedClickedListener) {
        setOnItemClickListener(addPodcastPresenter)
    }

    fun isPartiallyHidden(position: Int): Boolean {
        val isPartiallyHiddenBottom = (layoutManager as LinearLayoutManager).findLastCompletelyVisibleItemPosition() < position
        val isPartiallyHiddenTop = (layoutManager as LinearLayoutManager).findFirstCompletelyVisibleItemPosition() > position
        return isPartiallyHiddenBottom || isPartiallyHiddenTop
    }
}
